android/src/main/java/com/reactnativecommunity/webview/RNCWebViewManager.java

```diff
@@ -786,15 +791,52 @@ public void onPageStarted(WebView webView, String url, Bitmap favicon) {
 
     @Override
     public boolean shouldOverrideUrlLoading(WebView view, String url) {
-      progressChangedFilter.setWaitingForCommandLoadUrl(true);
-      dispatchEvent(
-        view,
-        new TopShouldStartLoadWithRequestEvent(
-          view.getId(),
-          createWebViewEvent(view, url)));
-      return true;
-    }
+      final RNCWebView rncWebView = (RNCWebView) view;
+      final JavaScriptContextHolder jsContext = ((ReactContext) view.getContext()).getJavaScriptContextHolder();
 
+      if (jsContext.get() != 0 && rncWebView.mCatalystInstance != null) {
+        final Pair<Integer, AtomicReference<RNCWebViewModule.ShouldOverrideUrlLoadingLock.ShouldOverrideCallbackState>> lock = RNCWebViewModule.shouldOverrideUrlLoadingLock.getNewLock();
+        final int lockIdentifier = lock.first;
+        final AtomicReference<RNCWebViewModule.ShouldOverrideUrlLoadingLock.ShouldOverrideCallbackState> lockObject = lock.second;
+
+        final WritableMap event = createWebViewEvent(view, url);
+        event.putInt("lockIdentifier", lockIdentifier);
+        rncWebView.sendDirectMessage("onShouldStartLoadWithRequest", event);
+
+        try {
+          assert lockObject != null;
+          synchronized (lockObject) {
+            final long startTime = System.nanoTime();
+            while (lockObject.get() == RNCWebViewModule.ShouldOverrideUrlLoadingLock.ShouldOverrideCallbackState.UNDECIDED) {
+              if (TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime) > 250) {
```

Comment:
any reason we need to do nano to milli conversion here? We could directly use System.currentTimeMillis() instead

Reply:
While the chances of a time change during this window seems very remote, it shouldn't be used for intervals as there's no guarantee that it won't jump forwards or backwards (which could result in a lot longer than 250ms!):

https://sangsoonam.github.io/2017/03/01/do-not-use-curenttimemillis-for-time-interval.html
https://mydailycodingexperience.blogspot.com/2016/08/systemcurrenttimemillis-vs.html
https://stackoverflow.com/questions/11954985/android-system-currenttimemillis-does-not-work-as-expected

So I guess I could use `SystemClock.elapsedRealtime()` instead.